tick/optim/solver/src/svrg.cpp

```diff
@@ -1,127 +1,150 @@
-            ArrayDouble iterate_no_interc = view(iterate, 0, n_features);
-            //
-            iterate_no_interc.mult_incr(x_i, delta);
-            iterate[n_features] += delta;
-            iterate.mult_incr(mu, -step);
-        } else {
-            iterate.mult_incr(x_i, delta);
-            iterate.mult_incr(mu, -step);
-        }
-
-        prox->call(iterate, step, iterate);
-
-        if (variance_reduction == VarianceReductionMethod::Random && t == rand_index)
-            next_iterate = iterate;
-
-        if (variance_reduction == VarianceReductionMethod::Average)
-            next_iterate.mult_incr(iterate, 1.0 / epoch_size);
+    if (variance_reduction == VarianceReductionMethod::Average) {
+      next_iterate.mult_incr(iterate, 1.0 / epoch_size);
     }
+  }
+  if (variance_reduction == VarianceReductionMethod::Last) {
+    next_iterate = iterate;
+  }
+  t += epoch_size;
+}
 
-    if (variance_reduction == VarianceReductionMethod::Last)
-        next_iterate = iterate;
+void SVRG::solve_sparse_proba_updates(bool use_intercept, ulong n_features) {
+  // Data is sparse, and we use the probabilistic update strategy
+  // This means that the model is a child of ModelGeneralizedLinear.
+  // The strategy used here uses non-delayed updates, with corrected
+  // step-sizes using a probabilistic approximation and the
+  // penalization trick: with such a model and prox, we can work only inside the current
+  // support (non-zero values) of the sampled vector of features
+  std::shared_ptr<ProxSeparable> casted_prox;
+  if (prox->is_separable()) {
+    casted_prox = std::static_pointer_cast<ProxSeparable>(prox);
+  }
+  for (t = 0; t < epoch_size; ++t) {
+    // Get next sample index
+    ulong i = get_next_i();
+    // Sparse features vector
+    BaseArrayDouble x_i = model->get_features(i);
+    // Gradients factors (model is a GLM)
+    // TODO: a grad_i_factor(i, array1, array2) to loop once on the features
+    double grad_i_diff = model->grad_i_factor(i, iterate) - model->grad_i_factor(i, fixed_w);
+    // We update the iterate within the support of the features vector, with the probabilistic correction
+    for (ulong idx_nnz = 0; idx_nnz < x_i.size_sparse(); ++idx_nnz) {
+      // Get the index of the idx-th sparse feature of x_i
+      ulong j = x_i.indices()[idx_nnz];
+      double full_gradient_j = full_gradient[j];
+      // Step-size correction for coordinate j
+      double step_correction = steps_correction[j];
+      // Gradient descent with probabilistic step-size correction
+      iterate[j] -= step * (x_i.data()[idx_nnz] * grad_i_diff + step_correction * full_gradient_j);
+      // Prox is separable, apply regularization on the current coordinate
+      iterate[j] = casted_prox->call_single(iterate[j], step * step_correction);
+    }
+    // And let's not forget to update the intercept as well
+    // It's updated at each step, so no step-correction and no prox applied it
+    if (use_intercept) {
```

Comment:
Why don't we apply call_single to it? If it's out of the range it should not be modified and if it is in (even if this is not the best idea ever) it would be consistent with the dense case.

Reply:
I'm not sure to understand. We can only use the call_single, that applies on a single coordinate, and this call_single does not use the range of the prox... Since we don't want to regularize the intercept, we just don't call call_single in this step.
Note that dense and sparse cases must match both with and without an intercept in the unit tests.

Reply:
Oh ok, what would happen if someone sets a prox to a defined range then? Won't we have different results in the dense and the sparse case?
Maybe call_single should take a (maybe optional) argument index that will determine if prox should be applied ot not?

Reply:
Ouch shit you're right, once again !